src/TaskLayout.ts

```diff
@@ -0,0 +1,72 @@
+/**
+ * Various rendering options for a query.
+ */
+export class LayoutOptions {
+    hideTaskCount: boolean = false;
+    hideBacklinks: boolean = false;
+    hidePriority: boolean = false;
+    hideStartDate: boolean = false;
+    hideScheduledDate: boolean = false;
+    hideDoneDate: boolean = false;
+    hideDueDate: boolean = false;
+    hideRecurrenceRule: boolean = false;
+    hideEditButton: boolean = false;
+    hideUrgency: boolean = true;
+    shortMode: boolean = false;
+}
+
+export type LayoutComponent =
+    | 'description'
+    | 'priority'
+    | 'recurrenceRule'
+    | 'startDate'
+    | 'scheduledDate'
+    | 'dueDate'
+    | 'doneDate'
+    | 'blockLink';
+
+/**
+ * This represents the desired layout of tasks when they are rendered in a given configuration.
+ * The layout is used when flattening the task to a string and when rendering queries, and can be
+ * modified by applying LayoutOptions objects.
+ */
+export class TaskLayout {
+    public defaultLayout: LayoutComponent[] = [
+        'description',
+        'priority',
+        'recurrenceRule',
+        'startDate',
+        'scheduledDate',
+        'dueDate',
+        'doneDate',
+        'blockLink',
+    ];
+    public layoutComponents: LayoutComponent[];
+    public options: LayoutOptions;
+
+    constructor(options?: LayoutOptions, components?: LayoutComponent[]) {
+        if (options) this.options = options;
+        else this.options = new LayoutOptions();
+        if (components) this.layoutComponents = components;
+        else this.layoutComponents = this.defaultLayout;
+
+        this.layoutComponents = this.applyOptions(this.options);
+    }
+
+    /**
+     * Return a new list of components with the given options applied.
+     */
+    applyOptions(layoutOptions: LayoutOptions): LayoutComponent[] {
+        const arrayRemove = (array: any[], value: any) => {
+            return array.filter((element) => element != value);
+        };
+        let newComponents = this.layoutComponents;
+        if (layoutOptions.hidePriority) newComponents = arrayRemove(newComponents, 'priority');
+        if (layoutOptions.hideRecurrenceRule) newComponents = arrayRemove(newComponents, 'recurrenceRule');
+        if (layoutOptions.hideStartDate) newComponents = arrayRemove(newComponents, 'startDate');
+        if (layoutOptions.hideScheduledDate) newComponents = arrayRemove(newComponents, 'scheduledDate');
+        if (layoutOptions.hideDueDate) newComponents = arrayRemove(newComponents, 'dueDate');
+        if (layoutOptions.hideDoneDate) newComponents = arrayRemove(newComponents, 'doneDate');
+        return newComponents;
```

Comment:
The pattern in names here is hard to see - for example to glance and check that the names are consistent on each line. I think it is because of the jagged/inconsistent alignment in the if condition.

I tried putting in {} and got this - which allows glancing down the `layout.hideBlah` lines and the option names very quickly.

```ts
        if (layoutOptions.hidePriority) {
            newComponents = arrayRemove(newComponents, 'priority');
        }
        if (layoutOptions.hideRecurrenceRule) {
            newComponents = arrayRemove(newComponents, 'recurrenceRule');
        }
        if (layoutOptions.hideStartDate) {
            newComponents = arrayRemove(newComponents, 'startDate');
        }
        if (layoutOptions.hideScheduledDate) {
            newComponents = arrayRemove(newComponents, 'scheduledDate');
        }
        if (layoutOptions.hideDueDate) {
            newComponents = arrayRemove(newComponents, 'dueDate');
        }
        if (layoutOptions.hideDoneDate) {
            newComponents = arrayRemove(newComponents, 'doneDate');
        }
```

Another option would be to move the if condition inside the local function, so it looked something approximately like:

```ts
newComponents = arrayRemove(newComponents, layoutOptions.hidePriority, 'priority');
newComponents = arrayRemove(newComponents, layoutOptions.hideRecurrenceRule, 'recurrenceRule');
newComponents = arrayRemove(newComponents, layoutOptions.hideStartDate, 'startDate');
newComponents = arrayRemove(newComponents, layoutOptions.hideScheduledDate, 'scheduledDate');
newComponents = arrayRemove(newComponents, layoutOptions.hideDueDate, 'dueDate');
newComponents = arrayRemove(newComponents, layoutOptions.hideDoneDate, 'doneDate');
```

It's probably obvious but I intentionally put the `layoutOptions.hide...` after the newComponents argument so the alignment aided readability.

Now any inconsistent values - like mixing due and start date on a line due to copying and pasting an existing line when new options are added in future, will jump out easily.

Reply:
Will fix.